Add docs for Dogu V3 artifacts (#92)
cloudogu/dogu-development-docs: docs/v3/concepts/artifacts_de.md

## Das Chart als Dogu-Deskriptor

Im Helm-Chart dient die `Chart.yaml` als führendes Manifest zur Bereitstellung zentraler Metadaten. Die Datei regelt sowohl die Identifikation des Charts als auch dessen Paket-Abhängigkeiten. Darüber hinaus existieren weitere Manifeste, die für die Integration in das CES eine entscheidende Rolle spielen. Nachfolgend werden die wichtigsten Artefakte und deren Bedeutung für ein Dogu V3 detailliert beschrieben. Zu den Top-Level-Dateien im Helm-Chart gehören:

```
k8s/
└─ helm/
   ├─ templates/
   │  └─ ...
   ├─ Chart.yaml
   ├─ chart-patch-tpl.yaml
   ├─ dogu-upgrade.yaml
   ├─ dogu-values-metadata.yaml
   ├─ values.yaml

>    └─ values.schema.json

Reviewer: Die Reihenfolge der Abschnitte je Artefakt unterscheidet sich von der hier genannten Reihenfolge. Sollten wir darauf eingehen? So etwa nach dem Motto "Diese Artefakte werden hier in Reihenfolge der Relevanz beschrieben."?

```

### `Chart.yaml`

In der `Chart.yaml` befinden sich Metadaten, die ein Dogu beschreiben. Im Chart werden zwei Versionen unterschieden. 
Die Dogu-Version entspricht der Version des Charts (`version`), während die App-Version (`appVersion`) die Version der fachlichen Anwendung innerhalb des Dogus beschreibt. Beide Versionen sind voneinander unabhängig und dürfen sich unterscheiden. So könnte beispielsweise Redmine als Dogu in Version `45.7.0` veröffentlicht werden, während die enthaltene Redmine-Anwendung die App-Version `6.1.2` besitzt. 

> Neben den [Helm-Standardfeldern](https://helm.sh/docs/topics/charts/#the-chartyaml-file) werden in der `Chart.yaml` plattformspezfische Annotations verwendet, die durch den Prefix `dogu.cloudogu.com/` gekenntzeichnet sind. Dieser Prefix kennzeichnet Dogu-Metadaten im Helm-Chart und grenzt sie von allgemeinen Kubernetes- oder Helm-Annotations ab. Folgende plattformspezische Annotations werden aktuell unterstützt:

Reviewer: typo:
```suggestion
Neben den [Helm-Standardfeldern](https://helm.sh/docs/topics/charts/#the-chartyaml-file) werden in der `Chart.yaml` plattformspezfische Annotations verwendet, die durch den Prefix `dogu.cloudogu.com/` gekennzeichnet sind. Dieser Prefix kennzeichnet Dogu-Metadaten im Helm-Chart und grenzt sie von allgemeinen Kubernetes- oder Helm-Annotations ab. Folgende plattformspezifische Annotations werden aktuell unterstützt:
```

- `api-version`: Die Dogu-API-Version, z.b. `v3`
- `display-name`: Der Anzeigename eines Dogus falls `name` Attribut der `Chart.yaml` nicht zur Anzeige verwendet werden soll
- `application.<name>`: Definiert weitere Anwendungen, die im Chart enthalten sind. `<name>` entspricht dem technischen Namen der enthaltenen Anwendung, wobei der Annotation-Wert eine nicht leere Versionsangabe als String enthält

> Unbekannte Annotations mit dem Prefix `dogu.cloudogu.com/` sind zulässig, werden jedoch von plattformspezischen Komponenten ignoriert, sofern sie nicht Teil der festgelegten Pflichtmetadaten sind. Als Pflichtfelder in der `Chart.yaml` gelten die Felder:

Reviewer: typo:

```suggestion
Unbekannte Annotations mit dem Prefix `dogu.cloudogu.com/` sind zulässig, werden jedoch von plattformspezifischen Komponenten ignoriert, sofern sie nicht Teil der festgelegten Pflichtmetadaten sind. Als Pflichtfelder in der `Chart.yaml` gelten die Felder:
```

annotations:
  dogu.cloudogu.com/api-version: v3
  dogu.cloudogu.com/display-name: Redmine
  dogu.cloudogu.com/application.redmine: "6.1.2"
  dogu.cloudogu.com/application.postgresql: "16.8"
```

### `values.yaml`

Die `values.yaml` dient als öffentliche Schnittstelle für die Konfiguration von Helm-Charts. Sie enthält Standardwerte, die für die Templates des Helm-Charts verwendet werden. Helm erlaubt es bei der Installation die Werte aus der `values.yaml` zu überschreiben, in dem es die übergebenen Werte mit den Standardwerten aus dem Helm-Chart zusammenführt. 

### `values.schema.json`

Für die `values.yaml` kann mithilfe der `values.schema.json` ein Schema in Form eines JSON-Schemas definiert werden, um die resultierenden Werte des Helm-Charts zu validieren. Da die Entwickler einer Anwendung am besten über den erforderlichen Input und dessen korrekte Struktur Bescheid wissen, sollte diese Schemadatei stets fester Bestandteil des Dogu-Helm-Charts sein. Der Dogu-Operator nutzt sie, um bei der Installation oder Aktualisierung eine automatisierte Validierung der Values durchzuführen und fehlerhafte Konfigurationen frühzeitig abzufangen.

### `dogu-values-metadata.yaml`

> Für die Plattform können globale Konfigurationparameter existieren, die nicht immer durch bestehende Werte aus der `values.yaml` eines Helm-Charts abgebildet werden können. So kann für die Plattform bespielsweise ein globales Log-Level gesetzt werden, welches für alle Dogus angewendet wird. Um diesem Umstand Rechnug zu tragen, wurde die `dogu-values-metadata.yaml` eingeführt. Durch sie können plattformspezifische Konfigurationen auf die Werte des Helm-Charts eines Dogus angewendet werden, indem in ihr das Mapping zwischen einzelnen Werten definiert wird. Plattformspezifische Konfigurationen werden in der Dogu-CR definiert und durch den Dogu-Operator angewendet. Folgend wird exemplarisch das Mapping des Log-Levels auf zwei Anwendungen eines Helm-Charts gezeigt:

Reviewer: Typos and a bit of shortening
```suggestion
Für die Plattform können globale Konfigurationsparameter existieren, die nicht immer durch bestehende Werte aus der `values.yaml` eines Helm-Charts abgebildet werden können. So kann für die Plattform bspw. ein globales Log-Level gesetzt werden, welches für alle Dogus angewendet wird. Um diesem Umstand Rechnung zu tragen, wurde die `dogu-values-metadata.yaml` eingeführt. Durch sie können plattformspezifische Konfigurationen auf die Werte des Helm-Charts eines Dogus angewendet werden, indem in ihr das Mapping zwischen einzelnen Werten definiert wird. Plattformspezifische Konfigurationen werden in der Dogu-CR definiert und durch den Dogu-Operator angewendet. Folgend wird exemplarisch das Mapping des Log-Levels auf zwei Anwendungen eines Helm-Charts gezeigt:
```

```yaml
apiVersion: v1
metavalues:

>   # Plattformspezische Konfigurationwert

Reviewer: Typo
```suggestion
  # Plattformspezifische Konfigurationwert
```

  mainLogLevel:
    keys:
      # ohne Mapping wird der Konfigurationswert ohne Änderung an die values.yaml durchgereicht 
      - path: controllerManager.env.loglevel
      # mit Mapping wird erst der Konfigurationswert abgebildet (z. B. panic -> error) und dann weitergereicht
      # Pfad zum Wert in der values.yaml für Hauptanwendung
      - path: app.env.loglevel
        mapping:
          DEBUG: debug
          INFO: info
          WARN: error
          ERROR: error
      # Pfad zum Wert in der values.yaml für zweite Anwendung
      - path: loglevel
        mapping:
          DEBUG: 1
          INFO: 2
          WARN: 3
          ERROR: 4      
```

**Dogu-CR, die den Konfigurationswert setzt:**

```yaml
apiVersion: k8s.cloudogu.com/v1
kind: Dogu
metadata:
  name: my-dogu
spec:
  name: official/my-dogu
  version: 1.2.3-4
  mappedValues:
    mainLogLevel: ERROR
```

> Um plattformspezische Konfigurationen treffen zu können, sollte ein Dogu-Helm-Chart die `dogu-values-metadata.yaml` stets mit ausliefern. Für konfigurationslose Dogus ist diese Datei optional. 

Reviewer: typo:

```suggestion
Um plattformspezifische Konfigurationen treffen zu können, sollte ein Dogu-Helm-Chart die `dogu-values-metadata.yaml` stets mit ausliefern. Für konfigurationslose Dogus ist diese Datei optional. 
```

## Templates und die Anbindung an die Plattform

In den Templates des Helm-Charts werden alle Kubernetes-Ressourcen zur Verfügung gestellt, um die Anwendung auf der Plattform ausführbar zu machen. Zu diesen Ressourcen zählen auch Custom Resources Definitions (CRDs), die von Cloudogu bereitgestellt werden und als API für die Plattform dienen. In diesem Abschnitt werden die einzelnen CRDs genauer beschrieben.

### AuthRegistration

Das CES bietet seinen Nutzer die Möglichkeit des Single Sign-On (SSO). Damit ein Dogu den SSO nutzen kann, muss es sich mithilfe der [`AuthRegistration-CR`](https://github.com/cloudogu/k8s-auth-registration-lib/blob/main/docs/operations/auth_registration_de.md) beim Identity-Provider (IdP) der Plattform registrieren. Aktuell werden die Authentifizierungsprotokolle `CAS`, `OAUTH` und `OIDC` unterstützt. Mit der Registration beim IdP werden zugleich Credentials in einem Secret bereitgestellt, die von dem Dogu für Anfragen gegen den IdP genutzt werden können. 

Eine `AuthRegistration` provisioniert nur die serverseitige Integration beim IdP. Die Anwendung muss das ausgewählte Authentifizierungsprotokoll weiterhin selbst implementieren.

### WarpMenuEntry

> Für die zentrale Navigation auf der Plattform wird das Warp-Menü verwendet. Wenn ein Dogu im Warp-Menü angezeigt werden soll, muss für jeden internen, für Anwender:innnen sichtbaren Einstiegspunkt eine [`WarpMenuEntry-CR`](https://github.com/cloudogu/k8s-warp-menu-entry-lib/blob/main/docs/operations/warp_menu_entry_de.md) bereitgestellt werden. Je nach Bedarf kann ein Dogu keinen, einen oder mehrere Einträge definieren.

Reviewer: lol too much
```suggestion
Für die zentrale Navigation auf der Plattform wird das Warp-Menü verwendet. Wenn ein Dogu im Warp-Menü angezeigt werden soll, muss für jeden internen, für Anwender:innen sichtbaren Einstiegspunkt eine [`WarpMenuEntry-CR`](https://github.com/cloudogu/k8s-warp-menu-entry-lib/blob/main/docs/operations/warp_menu_entry_de.md) bereitgestellt werden. Je nach Bedarf kann ein Dogu keinen, einen oder mehrere Einträge definieren.
```

Jeder Eintrag enthält einen deutschen und englischen Anzeigenamen, eine Kategorie sowie einen relativen Pfad zum Dogu. Das Warp-Menü macht die Anwendung nicht grundsätzlich erreichbar. Der relative Pfad muss zu einer unabhängig funktionierenden HTTP-Exposition passen.

### Exposition

Soll ein Dogu von außen erreichbar sein, muss hierfür eine oder mehrere [`Exposition-CR`](https://github.com/cloudogu/k8s-exposition-lib/blob/main/docs/operations/exposition_cr_de.md) deklariert werden. Die `Exposition` definiert, wie ein Service von außerhalb der Plattform erreichbar gemacht wird. Sie unterstützt HTTP-Routen (Layer 7) sowie rohe TCP- und UDP-Ports (Layer 4). Für HTTP-Routen können zusätzlich Path-Rewrites definiert werden. Auf Layer 4 Ebene kann es zu Kollisionen kommen, wenn zwei oder mehrere `Expositions` den gleichen Port für das gleiche Protokoll anfragen. In diesem Fall wird keine der betroffenen `Expositions` angewendet, solang der Konflikt besteht. Der Administrator wird hierüber im Status der CR informiert.  

### ServiceAccountRequest / ServiceAccountProducer

> Im CES (Cloudogu EcoSystem) ist es möglich, dass verschiedene Dogus miteinander interagieren. Hierfür muss für ein zugreifendes Dogu (Consumer) ein entsprechender ServiceAccount beim Ziel-Dogu (Producer) erstellt werden.

Reviewer: Already defined on top

Also, Compontents can likewise be producers/consumers

Also: Ring-Dependencies are possible because of the dependency-breaking handling:
```suggestion
Im CES ist es möglich, dass verschiedene Dogus/Components miteinander interagieren. Hierfür muss für ein zugreifendes Dogu/Component (Consumer) ein entsprechender ServiceAccount beim Ziel-Dogu/-Component (Producer) erstellt werden. Diese Service-Account-Beziehung ist also frei wählbar und ist nicht nur auf Dogus unter sich fest gelegt. Dieser Mechanismus erlaubt auch Ringabhängigkeiten zwischen Producer und Consumer.
```

Die ServiceAccount-Erstellung lässt sich deklarativ über die CRDs [`ServiceAccountRequest`](https://github.com/cloudogu/k8s-serviceaccount-lib/blob/main/docs/operations/serviceaccountrequest_cr_de.md) und [`ServiceAccountProducer`](https://github.com/cloudogu/k8s-serviceaccount-lib/blob/main/docs/operations/serviceaccountproducer_cr_de.md) steuern.

**`ServiceAccountProducer`**

> Bietet ein Dogu eine Schnittstelle an, die von anderen Dogus genutzt werden kann, muss es eine `ServiceAccountProducer-CR` bereitstellen. Die `ServiceAccountProducer` definiert, wie Service-Accounts für das Dogu erstellt werden und welche Parameter unterstützt werden. Ferner beschreibt die CR Werte, die das Dogu - der Producer - nach dem Erstellen eines Service-Accounts zurückgibt. Jeder zurückgegebene Wert wird als Schlüssel in das vom anfragenden Dogu - dem Consumer - referenzierte Secret geschrieben.

Reviewer: Zu spezifisch und gestörter Lesefluss
```suggestion
Bietet ein Producer eine Schnittstelle an, die von anderen Dogus genutzt werden kann, muss es eine `ServiceAccountProducer-CR` bereitstellen. Die `ServiceAccountProducer`-CR definiert, wie Service-Accounts für Consumer erstellt werden und welche Parameter unterstützt werden. Ferner beschreibt die CR die Struktur, wie die Werte nach dem Erstellen eines Service-Accounts vom Producer zurückgegeben werden. Jeder zurückgegebene Wert wird als Schlüssel in das vom anfragenden Consumer referenzierte Secret geschrieben.
```

**`ServiceAccountRequest`**

> Benötigt ein Dogu einen Service-Account bei einem anderen Dogu, muss es diesen über die `ServiceAccountRequest-CR` anfordern indem ein Producer benannt wird und optional Parameter übergeben werden. Die resultieren Credentials des Requests werden durch den Operator in ein referenziertes Kubernetes-Secret des konsumierenden Dogus geschrieben. Ist keine Referenz zu einem Secret im `ServiceAccountRequest` deklariert, wird ein Secret mit dem Name der ServiceAccountRequest-Ressource erstellt. Die CR bildet das Gegenstück zum `ServiceAccountProducer`. 

Reviewer: Zu spezifisch auf Dogu wie oben, Typo
```suggestion
Benötigt ein Consumer einen Service-Account bei einem Producer, muss es diesen über die `ServiceAccountRequest-CR` anfordern, indem ein Producer benannt wird und optional Parameter übergeben werden. Die resultieren Credentials des Requests werden durch den Operator in ein referenziertes Kubernetes-Secret des konsumierenden Dogus geschrieben. Ist keine Referenz zu einem Secret im `ServiceAccountRequest` deklariert, wird ein Secret mit dem Namen der ServiceAccountRequest-Ressource erstellt. Die CR bildet das Gegenstück zum `ServiceAccountProducer`. 
```
